### PR TITLE
[TACHYON-600] Add user model

### PR DESCRIPTION
**Pull request by dongche**
JIRA link: https://tachyon.atlassian.net/browse/TACHYON-600
This is a sub-task of TACHYON-593: Simple Authentication.

Add a User class for representing a user in Tachyon. It implements Principal interface in the context of Java security frameworks. Then it can be used in JAAS framework.

### CODE REVIEW

**Comment by haoyuan** [comment]
Jenkins, add to whitelist.

**Comment by haoyuan** [comment]
Jenkins, test this please.

File: common/src/main/java/tachyon/security/User.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import java.security.Principal;
+
+public class User implements Principal {
+  private String mName;
+
+  // TODO: add more attributes and methods for supporting Kerberos
+
+  public User(String mName) {
+    this.mName = mName;
```
Comment (calvinjia) [inline common/src/main/java/tachyon/security/User.java]: For Tachyon's code style, we avoid using `this`. Could you change the argument to something like `name`?

File: common/src/main/java/tachyon/security/User.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  @Override
+  public String getName() {
+    return mName;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
```
Comment (calvinjia) [inline common/src/main/java/tachyon/security/User.java]: Will this ever be true and `mName.equals(((User) o).mName)` false? If not, perhaps its clearer to skip this condition?

**Comment by dongche** [comment]
Updated. Thanks for your review! @calvinjia

File: common/src/main/java/tachyon/security/User.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.security;
+
+import java.security.Principal;
+
+public class User implements Principal {
+  private String mName;
```
Comment (calvinjia) [inline common/src/main/java/tachyon/security/User.java]: If the name cannot be set, perhaps define this as final?

**Comment by calvinjia** [comment]
Overall looks good, one other minor comment, thanks for the patch!

**Comment by ooq** [comment]
+1 on define mName as final, otherwise looks good to me. thanks!

File: common/src/main/java/tachyon/security/User.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.security;
+
+import java.security.Principal;
+
+public class User implements Principal {
```
Comment (apc999) [inline common/src/main/java/tachyon/security/User.java]: Could you javadoc this class very briefly in what it is? Even 1-2 sentences can be very helpful.

**Comment by apc999** [comment]
Overall looks good. What's the plan for unittest? Is it arriving as a separate PR? Alternatively, since this PR is relatively small, you can also include unittest file here.

**Comment by dongche** [comment]
Thanks for your comments! Patch updated with javadoc and a unit test.

**Comment by dongche** [comment]
@apc999 Thank you for changing the title. ^_^ 
I will pay attention to the pattern in following work.

**Comment by ooq** [comment]
Looks good to me. Thanks for the update.

**Comment by calvinjia** [comment]
+1 LGTM

File: common/src/test/java/tachyon/security/UserTest.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    Subject subject = new Subject();
+    subject.getPrincipals().add(new User("realUser"));
+    subject.getPrincipals().add(new User("proxyUser"));
+
+    // fetch added users
+    User[] users = subject.getPrincipals(User.class).toArray(new User[0]);
+
+    // verification
+    Assert.assertEquals("There should be 2 users.", 2, users.length);
+    Assert.assertFalse("The 2 users should be different.", users[0].equals(users[1]));
```
Comment (apc999) [inline common/src/test/java/tachyon/security/UserTest.java]: You don't need to put the `String` parameter in `Assert`.

Comment (apc999) [inline common/src/test/java/tachyon/security/UserTest.java]: For the second assert, use `Assert.assertNotEquals`?

Comment (dongche) [inline common/src/test/java/tachyon/security/UserTest.java]: Sure, sounds good. Thanks!
Updated in the latest patch.

**Comment by dongche** [comment]
Updated to address latest comments.

**Comment by apc999** [comment]
+1 LGTM! thanks for the patch!

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by haoyuan** [comment]
Jenkins, add to whitelist.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by haoyuan** [comment]
Thanks!

**Pull request merged**